### PR TITLE
[XPU] Use torch scaled_mm for XPU block FP8 linear

### PR DESCRIPTION
**Pull request by CaoE**
## Motivation

Enable the native Torch `scaled_mm` path for block-wise FP8 linear on Intel XPU.
The existing XPU path relies on the Triton block-FP8 implementation which is slower than `scaled_mm`.

## Modifications

- Use torch `scaled_mm` for block-wise FP8 linear instead of Triton block-FP8 implementation on XPU.
- Materialize transpose-contiguous block-scale storage once during `process_weights_after_loading_block_quant`. The post-load storage layout avoids repeated forward-time `contiguous/copy_` operations inside torch scaled_mm.

## Accuracy Tests

Tested on Intel BMG:
python -m pytest test/registered/e2e/xpu/test_xpu_fp8_linear.py

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #34953312376](https://github.com/sgl-project/sglang/actions/runs/34953312376)
Latest PR Test (Extra): :x: [Run #34953312321](https://github.com/sgl-project/sglang/actions/runs/34953312321)
Latest PR Test (AMD ROCm 10): :x: [Run #34953312587](https://github.com/sgl-project/sglang/actions/runs/34953312587)

### CODE REVIEW

File: python/sglang/srt/layers/quantization/fp8_utils.py [hunk 1]
```diff
@@ -262,6 +262,8 @@ def use_rowwise_torch_scaled_mm():
         # The condition is determined once as the operations
         # are time consuming.
         return get_device_capability() >= (9, 4) and torch_release >= (2, 7)
+    if _is_xpu:
+        return torch_release >= (2, 13)
```
Comment (polisettyvarma) [inline python/sglang/srt/layers/quantization/fp8_utils.py]: please confirm if it is 13 or 14

Comment (CaoE) [reply python/sglang/srt/layers/quantization/fp8_utils.py]: It is 13, and since torch is upgraded this can be removed, but it is ok to keep it for some specific branches. I'll update this PR today. Thanks.

File: python/sglang/srt/layers/quantization/fp8_utils.py [hunk 2]
```diff
@@ -563,6 +565,9 @@ def dispatch_w8a8_block_fp8_linear() -> Callable:
     """
     backend = get_fp8_gemm_runner_backend()
 
+    if _is_xpu:
+        return torch_w8a8_block_fp8_linear
```
Comment (polisettyvarma) [inline python/sglang/srt/layers/quantization/fp8_utils.py]: what happens on unsupported devices like bmg ?

File: python/sglang/srt/layers/quantization/fp8_utils.py [hunk 3]
```diff
@@ -563,6 +565,9 @@ def dispatch_w8a8_block_fp8_linear() -> Callable:
     """
     backend = get_fp8_gemm_runner_backend()
 
+    if _is_xpu:
```
Comment (rahulvijayaraghavan) [inline python/sglang/srt/layers/quantization/fp8_utils.py]: To support --fp8-gemm-backend=triton, check need to be below line 573

File: python/sglang/srt/layers/quantization/fp8_utils.py [hunk 4]
```diff
@@ -571,6 +576,42 @@ def dispatch_w8a8_block_fp8_linear() -> Callable:
...
+        q_input = input_2d
+        activation_scale = input_scale.view(-1, input_scale.shape[-1])
+
+    if not q_input.is_contiguous():
+        q_input = q_input.contiguous()
+    if not activation_scale.is_contiguous():
+        activation_scale = activation_scale.contiguous()
+    weight_t = weight.t()
+    weight_scale_t = weight_scale.t()
+    output = torch._scaled_mm(
```
Comment (rahulvijayaraghavan) [inline python/sglang/srt/layers/quantization/fp8_utils.py]: Is it better to use https://docs.pytorch.org/docs/2.13/generated/torch.nn.functional.scaled_mm.html

File: python/sglang/srt/layers/quantization/fp8_utils.py [hunk 5]
```diff
@@ -571,6 +576,42 @@ def dispatch_w8a8_block_fp8_linear() -> Callable:
...
+        q_input,
+        weight_t,
+        scale_a=activation_scale,
+        scale_b=weight_scale_t,
+        out_dtype=torch.bfloat16 if input_scale is not None else input.dtype,
+        bias=bias,
+    )
+    if isinstance(output, tuple):
```
Comment (rahulvijayaraghavan) [inline python/sglang/srt/layers/quantization/fp8_utils.py]: Check may not be required, since out is always Tensor

**Review by rahulvijayaraghavan — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/quantization/fp8_utils.py (hunk 1 above)
Comment (CaoE) [reply python/sglang/srt/layers/quantization/fp8_utils.py]: Updated

File: python/sglang/srt/layers/quantization/fp8_utils.py [hunk 6]
```diff
@@ -799,6 +854,8 @@ def _dispatch_auto_backend() -> Callable:
         )
 
         return npu_w8a8_mxfp8_linear
+    elif _is_xpu:
+        return torch_w8a8_block_fp8_linear
```
Comment (polisettyvarma) [inline python/sglang/srt/layers/quantization/fp8_utils.py:858]: will this work on bmg ?

File: python/sglang/srt/layers/quantization/fp8_utils.py (hunk 3 above)
Comment (CaoE) [reply python/sglang/srt/layers/quantization/fp8_utils.py]: Updated. dispatch for `torch_w8a8_block_fp8_linear` is moved to `_dispatch_auto_backend`.

**Review by polisettyvarma — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/quantization/fp8_utils.py (hunk 4 above)
Comment (CaoE) [reply python/sglang/srt/layers/quantization/fp8_utils.py]: Updated.

File: python/sglang/srt/layers/quantization/fp8_utils.py (hunk 5 above)
Comment (CaoE) [reply python/sglang/srt/layers/quantization/fp8_utils.py]: Updated.

File: python/sglang/srt/layers/quantization/fp8_utils.py (hunk 6 above)
Comment (CaoE) [reply python/sglang/srt/layers/quantization/fp8_utils.py:858]: Yes.

File: python/sglang/srt/layers/quantization/fp8_utils.py (hunk 2 above)
Comment (CaoE) [reply python/sglang/srt/layers/quantization/fp8_utils.py]: Updated. It is supported on BMG.

**Review by polisettyvarma — APPROVED** [review]

**Review by rahulvijayaraghavan — APPROVED** [review]

**Comment by CaoE** [comment]
@mingfeima Could you please review this PR? Thanks.

File: python/sglang/srt/layers/quantization/fp8_utils.py [hunk 7]
```diff
@@ -571,6 +573,58 @@ def dispatch_w8a8_block_fp8_linear() -> Callable:
...
+        q_input = q_input.contiguous()
+    if weight.stride(-1) != 1:
+        weight = weight.contiguous()
+    weight_t = weight.t()
+    scale_b = weight_scale if block_n == 1 else weight_scale.t()
+    output = torch.nn.functional.scaled_mm(
+        q_input,
+        weight_t,
+        activation_scale,
+        torch.nn.functional.ScalingType.BlockWise1x128,
```
Comment (polisettyvarma) [inline python/sglang/srt/layers/quantization/fp8_utils.py:619]: can we use 128x128 directly ?

Comment (CaoE) [reply python/sglang/srt/layers/quantization/fp8_utils.py:638]: The activation side follows SGLang's `1 x block_k` per-token-group convention. For the current regular block-FP8 path, block_k is 128, so activation uses `BlockWise1x128` with scale shape [M, K/128]. BlockWise128x128 applies to the weight operand after transposition, not to the activation. MXFP8 would use the separate 1 x 32 convention.

**Review by mingfeima — APPROVED** [review]

**Comment by mingfeima** [comment]
`block_n == 1` means weight_block_size=[1, 128] (weight scale [N, K/128]).
In SGLang that is not what current checkpoints use:

- Regular / DeepSeek-style block FP8 is [128, 128] (fp8.py, DeepSeek loaders, w4afp8).
- MXFP8 is [1, 32], which this function already rejects (block_k must be 128).
- Channel-wise FP8 goes through `apply_fp8_linear`, not `w8a8_block_fp8_linear`.

Activation scales: those are always per-token-group 1 x block_k (BlockWise1x128, shape [M, K/128]), even when the weight recipe is 128×128.

So leaving block_n = 1 is a guard from torch, but won't be called from sglang.

you can keep this PR unchanged.
